src/core/ComboBox/ComboBox.tsx

```diff
@@ -180,189 +174,274 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+        React.cloneElement(customItem, {
+          onClick: (e: unknown) => {
+            dispatch(['select', __originalIndex as number]);
+            customItem.props.onClick?.(e);
+          },
+          // ComboBox.MenuItem handles data-iui-index and iui-focused through context
+          // but we still need to pass them here for backwards compatibility with MenuItem
+          className: cx(customItem.props.className, 'iui-focused'),
+          'data-iui-index': __originalIndex,
+        })
+      ) : (
+        <ComboBoxMenuItem
+          key={option.id ?? __originalIndex}
+          {...rest}
+          isSelected={selectedIndex === __originalIndex}
+          onClick={() => dispatch(['select', __originalIndex])}
+          index={__originalIndex}
+        >
+          {option.label}
+        </ComboBoxMenuItem>
       );
-      switch (event.key) {
-        case 'ArrowDown':
-          if (isOpen) {
-            const nextIndex = Math.min(
-              focusedIndexInFilteredList + 1,
-              focusableOptions.length - 1,
-            );
-            setFocusedIndex(
-              options.findIndex(
-                (_, index) =>
-                  getOptionId(index) === focusableOptions[nextIndex].id,
-              ),
-            );
-          } else {
-            setIsOpen(true); // reopen menu if closed when typing
-          }
-          event.preventDefault();
-          event.stopPropagation();
-          break;
-        case 'ArrowUp':
-          if (isOpen) {
-            const previousIndex = Math.max(focusedIndexInFilteredList - 1, 0);
-            setFocusedIndex(
-              options.findIndex(
-                (_, index) =>
-                  getOptionId(index) === focusableOptions[previousIndex].id,
-              ),
-            );
-          }
-          event.preventDefault();
-          event.stopPropagation();
-          break;
-        case 'Enter':
-          if (isOpen) {
-            setSelectedValue(options[focusedIndex].value);
-            userOnChange.current?.(options[focusedIndex].value);
-          }
-          setIsOpen((open) => !open);
-          event.preventDefault();
-          event.stopPropagation();
-          break;
-        case 'Escape':
-          setIsOpen(false);
-          event.preventDefault();
-          event.stopPropagation();
-          break;
-        case 'Tab':
-          setIsOpen(false);
-          break;
-        default:
-          if (!isOpen) {
-            setIsOpen(true); // reopen menu if closed when typing
-          }
-          break;
-      }
     },
-    [focusedIndex, isOpen, options, getOptionId],
+    [focusedIndex, itemRenderer, memoizedItems, selectedIndex],
   );
 
-  const menuItems = React.useMemo(() => {
-    if (filteredOptions.length === 0) {
-      return [
-        <MenuExtraContent key={0}>
-          <Text isMuted>{emptyStateMessage}</Text>
-        </MenuExtraContent>,
-      ];
+  return (
+    <ComboBoxRefsContext.Provider
```

Comment:
Why 3 providers instead of 1?

Reply:
when any value inside a context changes, it rerenders the whole component that uses the context, even if that component is not using the part that changed. So it is better to split into multiple contexts.

Reply:
I know that it causes rerenders? But is it a case here?